### PR TITLE
Refactor Telemetry Tests

### PR DESCRIPTION
**Pull request by gargnupur**
[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[ ] Networking
[ ] Performance and Scalability
[ X] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

Pull Request Attributes

Please check any characteristics that apply to this pull request. 

[ ] Does not have any changes that may affect Istio users.

### CODE REVIEW

File: tests/integration/telemetry/stats/prometheus/nullvm/customize_metrics_test.go [hunk 1]
```diff
@@ -0,0 +1,129 @@
+// +build integ
```
Comment (bianpengyuan) [inline tests/integration/telemetry/stats/prometheus/nullvm/customize_metrics_test.go:1]: Is it possible to move the original test to here, so that update to it could shown clearly on github.

File: tests/integration/telemetry/stats/prometheus/nullvm/dashboard_test.go [hunk 2]
```diff
@@ -13,7 +13,7 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-package telemetry
+package nullvm
```
Comment (douglas-reid) [inline tests/integration/telemetry/stats/prometheus/nullvm/dashboard_test.go:16]: Is this the first part of a larger refactor ?  This test shouldn't really care about `nullvm` (and we want it to work across vms) right?

File: tests/integration/telemetry/stats/prometheus/nullvm/customize_metrics_test.go (hunk 1 above)
Comment (gargnupur) [reply tests/integration/telemetry/stats/prometheus/nullvm/customize_metrics_test.go:1]: I actually did the move like the dashboard test but somehow github took it as a copy paste :(

File: tests/integration/telemetry/stats/prometheus/nullvm/dashboard_test.go (hunk 2 above)
Comment (gargnupur) [reply tests/integration/telemetry/stats/prometheus/nullvm/dashboard_test.go:16]: Yes, want to see if this reduces runtime first and then lets discuss the exact directory structure we want to keep!

**Comment by gargnupur** [comment]
@howardjohn , @nmittler , @douglas-reid , @mandarjog : With a couple of changes in the tests, the test time is at around 30-35 mins for multicluster test suite currently. I feel that is doable to make them required?

Changes in this PR:
1) Move dashboard test as part of prometheus null vm setup
2) RequireSingleCluster for VM test
3) Make sending traffic and validation parallel in stackdriver and stats test. This really helps bring down test times from minutes to seconds.

Things left as per https://github.com/istio/istio/issues/26888#issuecomment-729916757: Moving customize metrics and zipkin also in one single suite deployment. Refactoring customize metrics test was leading to a lot of changes in basic stats test and hence I removed it from the above refactor. I feel we need a better way to apply data plane configs in the test with Istio deployed. Can we use istioctl operator files for that purpose?

File: tests/integration/telemetry/stats/prometheus/nullvm/dashboard_test.go (hunk 2 above)
Comment (gargnupur) [reply tests/integration/telemetry/stats/prometheus/nullvm/dashboard_test.go:16]: Doug, for a much larger refactor we need a better way of applying our configs..
For now I have just moved dashboard test. 

The nullvm in directory structure here refers to the fact that prometheus is not deployed using wasm vm. Should we rather call it default?

File: tests/integration/telemetry/stackdriver/stackdriver_filter_test.go [hunk 3]
```diff
@@ -96,28 +98,41 @@ func unmarshalFromTemplateFile(file string, out proto.Message, clName string) er
...
-					if err := validateTraces(t); err != nil {
-						return err
-					}
-					if err := validateEdges(t, clName); err != nil {
+
+			g, _ := errgroup.WithContext(context.Background())
+			for _, cltInstance := range clt {
+				g.Go(func() error {
+					err := retry.UntilSuccess(func() error {
+						if err := sendTraffic(t, cltInstance); err != nil {
```
Comment (nmittler) [inline tests/integration/telemetry/stackdriver/stackdriver_filter_test.go:106]: Why did you move sending traffic inside the loop? Shouldn't we just send traffic once and then wait for telemetry to show it?

File: tests/integration/telemetry/stackdriver/stackdriver_filter_test.go [hunk 4]
```diff
@@ -96,28 +98,41 @@ func unmarshalFromTemplateFile(file string, out proto.Message, clName string) er
...
+						}
+						if err := validateTraces(t); err != nil {
+							return err
+						}
+						if err := validateEdges(t, clName); err != nil {
+							return err
+						}
+						return nil
+
+					}, retry.Delay(3*time.Second), retry.Timeout(80*time.Second))
```
Comment (nmittler) [inline tests/integration/telemetry/stackdriver/stackdriver_filter_test.go:127]: The the timeout and delay ... should these values be used consistently for telemetry? And if so, should we make them flags so we can vary them based on platform (i.e. maybe we're running on real clusters vs kind)?

File: tests/integration/telemetry/stackdriver/stackdriver_filter_test.go (hunk 3 above)
Comment (howardjohn) [reply tests/integration/telemetry/stackdriver/stackdriver_filter_test.go:106]: The problem is if the traffic fails telemetry never shows up

Or in another case, it was applying config then sending requests immediately. if its too fast, the config never applied

Comment (gargnupur) [reply tests/integration/telemetry/stackdriver/stackdriver_filter_test.go:106]: I changed this behavior everywhere because many tests apply yaml files within the tests and it takes time for even deletion of that yaml to take effect...

File: tests/integration/telemetry/stats/prometheus/stats.go [hunk 5]
```diff
@@ -61,36 +60,60 @@ func GetPromInstance() prometheus.Instance {
...
-						return err
-					}
-					// This query will continue to increase due to readiness probe; don't wait for it to converge
-					if err := QueryFirstPrometheus(t, c, appQuery, GetPromInstance()); err != nil {
-						t.Logf("prometheus values for istio_echo_http_requests_total: \n%s", util.PromDump(c, promInst, "istio_echo_http_requests_total"))
+			g, _ := errgroup.WithContext(context.Background())
+			for _, cltInstance := range client {
+				g.Go(func() error {
+					err := retry.UntilSuccess(func() error {
+						if err := SendTraffic(t, cltInstance); err != nil {
```
Comment (nmittler) [inline tests/integration/telemetry/stats/prometheus/stats.go:88]: same question here.

File: tests/integration/telemetry/stackdriver/stackdriver_filter_test.go (hunk 4 above)
Comment (gargnupur) [reply tests/integration/telemetry/stackdriver/stackdriver_filter_test.go:127]: They at times depend upon the test...

@douglas-reid : what do you think, should we make it a constant?

File: tests/integration/telemetry/stats/prometheus/stats.go (hunk 5 above)
Comment (gargnupur) [reply tests/integration/telemetry/stats/prometheus/stats.go:88]: replied above..

File: tests/integration/telemetry/stackdriver/stackdriver_filter_test.go (hunk 4 above)
Comment (douglas-reid) [reply tests/integration/telemetry/stackdriver/stackdriver_filter_test.go:127]: const sounds reasonable to me. i don't think these should vary by platform too much, unless a platform is _really_ CPU constrained or something.  even then, 80 seconds is a long time for retrying

File: tests/integration/telemetry/stats/prometheus/nullvm/dashboard_test.go (hunk 2 above)
Comment (douglas-reid) [reply tests/integration/telemetry/stats/prometheus/nullvm/dashboard_test.go:16]: i think nullvm is fine for now.

File: tests/integration/telemetry/stackdriver/stackdriver_filter_test.go (hunk 4 above)
Comment (gargnupur) [reply tests/integration/telemetry/stackdriver/stackdriver_filter_test.go]: Thanks..Fixed

File: tests/integration/telemetry/constants.go [hunk 6]
```diff
@@ -19,4 +19,8 @@ const (
 	// For multicluster tests we multiply the number of requests with a
 	// constant multiplier to make sure we have cross cluster traffic.
 	RequestCountMultipler = 3
+	// Default retry delay used in tests.
+	RetryDelayInSeconds = 3
```
Comment (nmittler) [inline tests/integration/telemetry/constants.go:23]: Even better, could we make these flags with the values here as the defaults? That way we can tweak them dynamically if needed.

Comment (gargnupur) [reply tests/integration/telemetry/constants.go]: Good idea.. done

File: tests/integration/telemetry/flags.go [hunk 7]
```diff
@@ -16,30 +16,18 @@
...
-		Setup(istio.Setup(&i, nil)).
-		Setup(func(ctx resource.Context) (err error) {
-			for _, cl := range ctx.Clusters() {
-				ingr = append(ingr, i.IngressFor(cl))
-			}
-			return nil
-		}).
-		Run()
+func init() {
+	flag.DurationVar(&RetryDelay, "istio.test.telemetry.retryDelay", time.Second*3, "Default retry delay used in tests")
```
Comment (nmittler) [inline tests/integration/telemetry/flags.go:31]: I think you also have to update https://github.com/istio/istio/blob/master/pkg/test/framework/initstatics.go to import this package so that the flags are registered.

**Review by nmittler — APPROVED** [review]

File: pkg/test/framework/components/telemetry/flags.go [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+
+var (
+	// RetryDelay is the retry delay used in tests.
+	RetryDelay time.Duration
+	// RetryTimeout is the retry timeout used in tests.
+	RetryTimeout time.Duration
+)
+
+func init() {
+	flag.DurationVar(&RetryDelay, "istio.test.telemetry.retryDelay", time.Second*3, "Default retry delay used in tests")
```
Comment (douglas-reid) [inline pkg/test/framework/components/telemetry/flags.go:30]: are these used in all telemetry tests now? if not, do we have a way to track full updating of all tests to use these flags?

**Review by douglas-reid — APPROVED** [review]

File: pkg/test/framework/components/telemetry/flags.go (hunk 8 above)
Comment (gargnupur) [reply pkg/test/framework/components/telemetry/flags.go:30]: Added...https://github.com/istio/istio/issues/29467
